monai/utils/sliding_window_inference.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import math
+import numpy as np
+import torch
+from monai.data.transforms import ImageEndPadder
+
+
+def sliding_window_inference(inputs, roi_size, sw_batch_size, predictor, device):
+    """Use SlidingWindow method to execute inference.
+
+    Args:
+        roi_size (list, tuple): the window size to execute SlidingWindow inference.
+        sw_batch_size (int): the batch size to run window slices.
+        predictor: a portion from numpy.lib.arraypad.pad is copied below.
+        device: on which device to execute model inference, cpu or gpu.
+
+    Note:
+        must be channel first, support both 2D and 3D.
+        input data must have batch dim.
+        execute on 1 image/per inference, run a batch of window slices of 1 input image.
+    """
+    num_spatial_dims = len(inputs.shape) - 2
+    assert len(roi_size) == num_spatial_dims, 'roi_size {} does not match input dims.'.format(roi_size)
+
+    # determine image spatial size and batch size
+    # Note: all input images must have the same image size and batch size
+    image_size = list(inputs.shape[2:])
+    batch_size = inputs.shape[0]
+
+    # TODO: Enable batch sizes > 1 in future.
+    assert batch_size == 1, "input batch size must be 1"
+
+    # in case that image size is smaller than roi size
+    is_oversized = False
+    original_image_size = [image_size[i] for i in range(num_spatial_dims)]
+
+    for i in range(num_spatial_dims):
+        if int(roi_size[i]) > image_size[i]:
+            is_oversized = True
+            break
+
+    if is_oversized:
+        for i in range(num_spatial_dims):
+            image_size[i] = max(image_size[i], roi_size[i])
+
+        padder = ImageEndPadder(roi_size, 'constant')
+        inputs = padder(inputs)
+
+    scan_interval = _get_scan_interval(image_size, roi_size, num_spatial_dims)
+    scan_num = [int(math.ceil(float(image_size[i]) / scan_interval[i])) for i in range(num_spatial_dims)]
+
+    # Store all slices in list.
+    slices = []
+    if num_spatial_dims == 3:
```

Comment:
This whole section could perhaps be done with `utils.arrayutils.iter_patch_slices`.

Reply:
Hi @ericspod ,

The reason I kept this logic here is that: Wenqi told me we may need to support overlap between window slices later. So maybe we can refactor in future sprint when more requirements come?
Thanks.

Reply:
Hi @Nic-Ma It makes sense to put it in utils. It will enhance readability and modularity. If we need to support slices overlap, we can create more functions/refactor then.